pkg/tanka/find.go

```diff
@@ -2,123 +2,165 @@ package tanka
-	envs, errs := find(path, Opts{JsonnetOpts: opts.JsonnetOpts})
-	if errs != nil {
-		return envs, ErrParallel{errors: errs}
-	}
-
-	// optionally filter
-	if opts.Selector == nil || opts.Selector.Empty() {
-		return envs, nil
-	}
+	return findEnvsFromPaths([]string{path}, opts)
+}
 
-	filtered := make([]*v1alpha1.Environment, 0, len(envs))
-	for _, e := range envs {
-		if !opts.Selector.Matches(e.Metadata) {
-			continue
-		}
-		filtered = append(filtered, e)
-	}
+// FindEnvsFromPaths does the same as FindEnvs but takes a list of paths instead
+func FindEnvsFromPaths(paths []string, opts FindOpts) ([]*v1alpha1.Environment, error) {
+	return findEnvsFromPaths(paths, opts)
+}
 
-	return filtered, nil
+type findJsonnetFilesOut struct {
+	jsonnetFiles []string
+	err          error
 }
 
-func findErr(path string, err error) []error {
-	return []error{fmt.Errorf("%s:\n %w", path, err)}
+type findEnvsOut struct {
+	envs []*v1alpha1.Environment
+	err  error
 }
 
-// find implements the actual functionality described at 'FindEnvs'
-func find(path string, opts Opts) ([]*v1alpha1.Environment, []error) {
-	// try if this has envs
-	list, err := List(path, opts)
-	if err != nil &&
-		// expected when looking for environments
-		!errors.As(err, &jpath.ErrorNoBase{}) &&
-		!errors.As(err, &jpath.ErrorFileNotFound{}) {
-		return nil, findErr(path, err)
-	}
-	if len(list) != 0 {
-		// it has. don't search deeper
-		return list, nil
+func findEnvsFromPaths(paths []string, opts FindOpts) ([]*v1alpha1.Environment, error) {
+	if opts.Parallelism >= 0 {
+		opts.Parallelism = runtime.NumCPU()
 	}
 
-	stat, err := os.Stat(path)
-	if err != nil {
-		return nil, findErr(path, err)
+	log.Debug().Int("parallelism", opts.Parallelism).Int("paths", len(paths)).Msg("Finding Tanka environments")
+	startTime := time.Now()
+
+	// find all jsonnet files within given paths
+	pathChan := make(chan string, len(paths))
+	findJsonnetFilesChan := make(chan findJsonnetFilesOut)
+	for i := 0; i < opts.Parallelism; i++ {
+		go func() {
+			for path := range pathChan {
+				jsonnetFiles, err := jsonnet.FindFiles(path, nil)
+				var mainFiles []string
+				for _, file := range jsonnetFiles {
+					if filepath.Base(file) == jpath.DefaultEntrypoint {
+						mainFiles = append(mainFiles, file)
+					}
+				}
+				findJsonnetFilesChan <- findJsonnetFilesOut{jsonnetFiles: mainFiles, err: err}
+			}
+		}()
 	}
 
-	// if path is a file, don't search deeper
-	if !stat.IsDir() {
-		return nil, nil
+	// push paths to channel
+	var pathMap = map[string]bool{} // prevent duplicates
+	for _, path := range paths {
+		if _, ok := pathMap[path]; ok {
+			continue
+		}
+		pathMap[path] = true
+		pathChan <- path
 	}
 
-	// list directory
-	files, err := os.ReadDir(path)
-	if err != nil {
-		return nil, findErr(path, err)
+	// collect jsonnet files
+	var jsonnetFiles []string
+	for i := 0; i < len(paths); i++ {
+		res := <-findJsonnetFilesChan
+		if res.err != nil {
+			return nil, res.err
+		}
```

Comment:
This makes me think you could use a `sync.errgroup` here. Those have direct error support. As a bonus we could arrange to pass a context in and then get nicer cancellation behaviour.

Reply:
I actually converted it to return a list of errors as well